Include/pymath.h

```diff
@@ -221,10 +221,27 @@ PyAPI_FUNC(void) _Py_set_387controlword(unsigned short);
 #define _Py_IntegralTypeMax(type) ((_Py_IntegralTypeSigned(type)) ? (((((type)1 << (sizeof(type)*CHAR_BIT - 2)) - 1) << 1) + 1) : ~(type)0)
 /* Return the minimum value of integral type *type*. */
 #define _Py_IntegralTypeMin(type) ((_Py_IntegralTypeSigned(type)) ? -_Py_IntegralTypeMax(type) - 1 : 0)
-/* Check whether *v* is in the range of integral type *type*. This is most
- * useful if *v* is floating-point, since demoting a floating-point *v* to an
- * integral type that cannot represent *v*'s integral part is undefined
- * behavior. */
-#define _Py_InIntegralTypeRange(type, v) (_Py_IntegralTypeMin(type) <= v && v <= _Py_IntegralTypeMax(type))
+
+/* Check if the floating-point number v (double) would overflow when casted to
+ * the integral type 'type'.
+ *
+ * Test (double)type_min(type) <= v <= (double)type_max(type) where v is a
+ * double, and type_min() and type_max() integers are rounded towards zero when
+ * casted to a double.
+ *
+ * (double)int cast rounds to nearest with ties going to nearest even integer
+ * (ROUND_HALF_EVEN). Use nextafter() to round towards zeros (ROUND_DOWN).
+ *
+ * For example, _Py_IntegralTypeMax(int64_t)=2**63-1 casted to double gives
+ * 2**63 which is greater than 2**63-1. The problem is that "v <= 2**63" fails
+ * to detect that v will overflow when casted to int64_t.
+ * nextafter((double)(2**63-1), 0.0) gives the floating-point number 2**63-1024
+ * which is less than or equal to the integer 2**63-1 and so can be used to
+ * test that v would overflow.
+ *
+ * In short, nextafter((double)x, 0.0) rounds the integer x towards zero. */
+#define _Py_DoubleInIntegralTypeRange(type, v) \
+    (nextafter((double)_Py_IntegralTypeMin(type), 0.0) <= v \
```

Comment:
Sorry, but this isn't right.

For a 32-bit signed integer type, this condition becomes:
```
-2147483647.9999998 <= v <= 2147483646.9999998
```

but what we actually want is:
```
-2147483649.0 < v < 2147483648.0
```

For a 64-bit signed integer type, this condition becomes:
```
-9223372036854774784.0 <= v <= 9223372036854774784.0
```

but the actual limits we need are
```
-9223372036854775808.0 <= v <= 9223372036854774784.0
```

Reply:
Suggested alternative approach: given a double `v` that we want to safely convert to an `int` (say), first use `modf` to extract the fractional part `fv` and the integer part `iv` of `v` (both `fv` and `iv` are still doubles at this point). We can just ignore the fractional part `fv`: it doesn't affect whether we can safely cast to `int` or not. Now for `iv`, check whether it satisfies `(double)INT_MIN <= iv` and `iv < -(double)INT_MIN`. Note that `INT_MIN` is (always in practice, even though not guaranteed by the standard) a power of two, so the conversion to `double` doesn't change the value and we're effectively checking whether `iv` lies in the half-open interval `[INT_MIN, -INT_MIN)`.  Since `iv` is an integer, that's equivalent to checking whether `iv` lies in the closed interval `[INT_MIN, -INT_MIN-1]`. But making all our usual assumptions about integer representation (two's complement, no trap representation, no padding bits, etc.), `-INT_MIN-1` is the same as `INT_MAX`.

Similarly for `long`, `long long`, etc.